packages/flutter/lib/src/services/platform_messages.dart

```diff
@@ -237,3 +250,647 @@ class PlatformMessages {
+        break;
+      case _kFloat64:
+        result = buffer.getFloat64();
+        break;
+      case _kString:
+        final int length = _readSize(buffer);
+        result = UTF8.decoder.convert(buffer.getUint8List(length));
+        break;
+      case _kUint8List:
+        final int length = _readSize(buffer);
+        result = buffer.getUint8List(length);
+        break;
+      case _kInt32List:
+        final int length = _readSize(buffer);
+        result = buffer.getInt32List(length);
+        break;
+      case _kInt64List:
+        final int length = _readSize(buffer);
+        result = buffer.getInt64List(length);
+        break;
+      case _kFloat64List:
+        final int length = _readSize(buffer);
+        result = buffer.getFloat64List(length);
+        break;
+      case _kList:
+        final int length = _readSize(buffer);
+        result = new List<dynamic>(length);
+        for (int i = 0; i < length; i++) {
+          result[i] = _readValue(buffer);
+        }
+        break;
+      case _kMap:
+        final int length = _readSize(buffer);
+        result = new Map<dynamic, dynamic>();
+        for (int i = 0; i < length; i++) {
+          result[_readValue(buffer)] = _readValue(buffer);
+        }
+        break;
+      default: throw new FormatException('Message corrupted');
+    }
+    return result;
+  }
+
+  @override
+  ByteData encodeMethodCall(String name, dynamic arguments) {
+    assert(name != null);
+    final _WriteBuffer buffer = new _WriteBuffer();
+    _writeValue(buffer, name);
+    _writeValue(buffer, arguments);
+    return buffer.byteData;
+  }
+
+  @override
+  dynamic decodeEnvelope(ByteData envelope) {
+    // First byte is zero in success case, and non-zero otherwise.
+    if (envelope == null || envelope.lengthInBytes == 0)
+      throw new FormatException('Expected envelope, got nothing');
+    final _ReadBuffer buffer = new _ReadBuffer(envelope);
+    if (buffer.getUint8() == 0)
+      return _readValue(buffer);
+    final dynamic errorCode = _readValue(buffer);
+    final dynamic errorMessage = _readValue(buffer);
+    final dynamic errorDetails = _readValue(buffer);
+    if (errorCode is String && (errorMessage == null || errorMessage is String))
+      throw new PlatformException(code: errorCode, message: errorMessage, details: errorDetails);
+    else
+      throw new FormatException('Invalid envelope');
+  }
+}
+
+class _WriteBuffer {
+  final Uint8Buffer buffer = new Uint8Buffer();
+
+  ByteData _fourBytes;
+  Uint8List _fourBytesAsList;
+  ByteData _eightBytes;
+  Uint8List _eightBytesAsList;
+
+  _WriteBuffer() {
+    _fourBytes = new ByteData(4);
+    _fourBytesAsList = _fourBytes.buffer.asUint8List();
+    _eightBytes = new ByteData(8);
+    _eightBytesAsList = _eightBytes.buffer.asUint8List();
+  }
+
+  void putUint8(int byte) {
+    buffer.add(byte);
+  }
+
+  void putInt32(int value) {
+    _fourBytes.setInt32(0, value);
+    buffer.addAll(_fourBytesAsList);
```

Comment:
Direct bit-smashing applied for ints.

For doubles, it seems I would need to create a new ByteData view instance of buffer.buffer for each double added, because the buffer is dynamically extended. Retained original approach for now.